src/coreclr/src/inc/crosscomp.h

```diff
@@ -393,6 +393,62 @@ typedef struct _T_KNONVOLATILE_CONTEXT_POINTERS {
 
 #endif
 
+#if defined(DAC_COMPILE) && defined(TARGET_UNIX)
+// This is a TARGET oriented copy of CRITICAL_SECTION and PAL_CS_NATIVE_DATA_SIZE
+// It is configured based on TARGET configuration rather than HOST configuration
+// There is validation code in src/coreclr/src/vm/crst.cpp to keep these from
+// getting out of sync
+
+#define T_CRITICAL_SECTION_VALIDATION_MESSAGE "T_CRITICAL_SECTION validation failed. It is not in sync with CRITICAL_SECTION"
+
+#if defined(TARGET_DARWIN) && defined(TARGET_X86)
+#define DAC_CS_NATIVE_DATA_SIZE 76
+#elif defined(TARGET_DARWIN) && defined(TARGET_AMD64)
+#define DAC_CS_NATIVE_DATA_SIZE 120
+#elif defined(TARGET_FREEBSD) && defined(TARGET_X86)
+#define DAC_CS_NATIVE_DATA_SIZE 12
+#elif defined(TARGET_FREEBSD) && defined(TARGET_AMD64)
+#define DAC_CS_NATIVE_DATA_SIZE 24
+#elif defined(TARGET_LINUX) && defined(TARGET_ARM)
+#define DAC_CS_NATIVE_DATA_SIZE 80
+#elif defined(TARGET_LINUX) && defined(TARGET_ARM64)
+#define DAC_CS_NATIVE_DATA_SIZE 116
+#elif defined(TARGET_LINUX) && defined(TARGET_X86)
+#define DAC_CS_NATIVE_DATA_SIZE 76
+#elif defined(TARGET_LINUX) && defined(TARGET_AMD64)
+#define DAC_CS_NATIVE_DATA_SIZE 96
+#elif defined(TARGET_NETBSD) && defined(TARGET_AMD64)
+#define DAC_CS_NATIVE_DATA_SIZE 96
+#elif defined(TARGET_NETBSD) && defined(TARGET_ARM)
+#define DAC_CS_NATIVE_DATA_SIZE 56
+#elif defined(TARGET_NETBSD) && defined(TARGET_X86)
+#define DAC_CS_NATIVE_DATA_SIZE 56
+#else
+#warning
+#error  DAC_CS_NATIVE_DATA_SIZE is not defined for this architecture
```

Comment:
@sdmaclea, could you please specify which struct this size correspond to?
I have ported `PAL_CS_NATIVE_DATA_SIZE` in `pal.h` to Solaris and it was clear that we are dealing with size of `struct PAL_CS_NATIVE_DATA` (which in case of solaris x64 is 48), but here I couldn't find `struct DAC_CS_NATIVE_DATA`. If it is the size of `struct T_CRITICAL_SECTION`, should we consider renaming it to `T_CRITICAL_SECTION_SIZE` to make it self-explanatory?

Reply:
See https://github.com/dotnet/runtime/search?q=PAL_CS_NATIVE_DATA_SIZE&unscoped_q=PAL_CS_NATIVE_DATA_SIZE

It is the DAC's copy of `PAL_CS_NATIVE_DATA_SIZE`.  You can see it in the assert in src/coreclr/src/vm/crst.cpp. They are supposed to be the same value.

If you want to add a comment to that effect, it seems like a good idea. I assume this could easily go in your solaris PR. If you would rather I do it, I can create a PR.

Reply:
Thank you for the information. I will add it to my Solaris port branch. I was thinking to update the error message to the effect “what to look for”, as this error case is to mainly assist folks porting code to new platforms. I still have some code base to cover, so it may take few days for me to prepare a PR.